infrastructure/src/main/java/org/corfudb/infrastructure/redundancy/RedundancyCalculator.java

```diff
@@ -0,0 +1,394 @@
+                .collect(ImmutableList.toImmutableList());
+
+        ImmutableList<AgedSegment> newSegments = newList
+                .stream()
+                .map(segment -> new AgedSegment(segment, CURRENT_EPOCH))
+                .collect(ImmutableList.toImmutableList());
+
+        ImmutableList<AgedSegment> allSegments = new ImmutableList.Builder<AgedSegment>()
+                .addAll(oldSegments)
+                .addAll(newSegments)
+                .build()
+                .stream()
+                .sorted()
+                .collect(ImmutableList.toImmutableList());
+
+        ImmutableList<AgedSegment> initList = ImmutableList.of();
+
+        // Merge the old and the new segments and return the new list.
+        return allSegments.stream()
+                .reduce(initList,
+                        (accumulatedList, nextSegment) ->
+                                mergeSegmentToAccumulatedList(nextSegment, accumulatedList),
+                        (list1, list2) -> list2)
+                .stream()
+                .map(agedSegment -> agedSegment.segment)
+                .collect(ImmutableList.toImmutableList());
+    }
+
+    /**
+     * Given a layout, and global trim mark creates an initial list
+     * of non-empty and bounded transfer segments.
+     *
+     * @param layout   A current layout.
+     * @param trimMark A current global trim mark.
+     * @return A list of transfer segments.
+     */
+    public ImmutableList<TransferSegment> createStateList(Layout layout, long trimMark) {
+        return layout.getSegments()
+                .stream()
+                // Keep all the segments after the trim mark, except the open one.
+                .filter(segment -> segment.getEnd() != NON_ADDRESS && segment.getEnd() >= trimMark)
+                .map(segment -> {
+                    // The transfer segment's start is the layout segment's start or a trim mark,
+                    // whichever is greater.
+                    long segmentStart = Math.max(segment.getStart(), trimMark);
+                    // The transfer segment's end should be inclusive.
+                    // It is the last address to transfer.
+                    long segmentEnd = segment.getEnd() - 1L;
+
+                    if (segmentContainsServer(segment, getServer())) {
+                        TransferSegmentStatus restored = TransferSegmentStatus
+                                .builder()
+                                .segmentState(RESTORED)
+                                .totalTransferred(segmentEnd - segmentStart + 1L)
+                                .build();
+
+                        return TransferSegment
+                                .builder()
+                                .startAddress(segmentStart)
+                                .endAddress(segmentEnd)
+                                .status(restored)
+                                .build();
```

Comment:
It's important to keep track of them in a state list. In the next PR with an asynchronous transfer, it will be important since we need to keep track of the status of every segment with higher granularity and make sure all of them are restored at the end.